src/sentry/debug_files/artifact_bundle_indexing.py

```diff
@@ -227,70 +171,146 @@ def get_flat_file_meta(self) -> Optional[FlatFileMeta]:
+                dist_name=identifier.dist,
+            )
+            FlatFileIndexState.objects.update_or_create(
+                flat_file_index=flat_file_index,
+                artifact_bundle=artifact_bundle,
+                defaults={
+                    "indexing_state": ArtifactBundleIndexingState.NOT_INDEXED.value,
+                    "date_added": timezone.now(),
+                },
+            )
+
+    return identifiers
 
-        return locks.get(locking_key, duration=60 * 10, name="bundle_index")
+
+@sentry_sdk.tracing.trace
+def remove_artifact_bundle_from_indexes(artifact_bundle: ArtifactBundle):
+    flat_file_indexes = ArtifactBundleFlatFileIndex.objects.filter(
+        flatfileindexstate__artifact_bundle=artifact_bundle
+    )
+    for idx in flat_file_indexes:
+        identifier = FlatFileIdentifier(
+            project_id=idx.project_id, release=idx.release_name, dist=idx.dist_name
+        )
+
+        was_removed = update_artifact_bundle_index(
+            identifier, bundles_to_remove=[artifact_bundle.id]
+        )
+        if not was_removed:
+            metrics.incr("artifact_bundle_flat_file_indexing.removal.would_block")
+            # TODO: mark this for async removal
+            pass
 
 
 @sentry_sdk.tracing.trace
 def update_artifact_bundle_index(
-    bundle_meta: BundleMeta, bundle_archive: ArtifactBundleArchive, identifier: FlatFileIdentifier
-):
+    identifier: FlatFileIdentifier,
+    blocking: bool = False,
+    bundles_to_add: List[BundleManifest] | None = None,
+    bundles_to_remove: List[int] | None = None,
+) -> bool:
     """
-    This will merge the `ArtifactBundle` given via `bundle_meta` and `bundle_archive`
-    into the index identified via `identifier`.
+    This will update the index identified via `identifier`.
+    Multiple manifests given in `bundles_to_add` and `bundles_to_remove` will be merged
+    into the index as one batched operation.
 
-    If this function fails for any reason, it can be, and *has to be* retried at a later point.
+    If this function fails for any reason, it can be, and *has to be* retried at a later point,
+    as not doing so will leave inconsistent indexes around.
     """
-    # TODO: maybe query `FlatFileIndexState` to avoid double-indexing?
-
-    lock = identifier.get_lock()
-    with TimedRetryPolicy(60)(lock.acquire):
-        flat_file_index = ArtifactBundleFlatFileIndex.objects.filter(
-            project_id=identifier.project_id,
-            release_name=identifier.release,
-            dist_name=identifier.dist,
-        ).first()
+    with atomic_transaction(
+        using=(
+            router.db_for_write(ArtifactBundleFlatFileIndex),
+            router.db_for_write(FlatFileIndexState),
+        )
+    ):
+        # The `nowait=True` will opportunistically lock the row/index without blocking,
+        # and throw an error otherwise which we will pass down and handle in the caller.
+        try:
+            flat_file_index = (
```

Comment:
How long will the lock be kept? Only during read?

Reply:
If so, we might have problems when two concurrent tasks read the same `.json` index in sequence and they update it. This is because the merging function is commutative and idempotent but the actual write in nodestore doesn't know anything about changes and only applies updates in a last writer wins fashion.

Reply:
the `FOR UPDATE` lock should be held until the end of the transaction, or at least until the `update_flat_file_index` which does an `UPDATE` internally.

Reply:
Perfect! Then my doubts are clarified! Thanks